feat: track exceptions in :async activejob adapter

halfbyte opened the pull request:
Fixes #361

I have some doubts about this but I wanted to give it a shot.

- I am not 100% sure if the way this is installed will work with all of the various loader configurations
- I am not sure if I have set up the call to `.notify` correctly and if my usage of `.flush` is correct.
- ~~I am not 100% sure if this would survive a user installing additional callbacks (I will try to add a test for that)~~
- Due to the way ActiveJob testing works, I'm not actually testing the `:async` adapter (I don't think that that's a huge problem as I am not monkeypatching anything)

Please have a look.

joshuap commented on spec/integration/rails/async_queue_adapter_spec.rb:
```diff
@@ -0,0 +1,18 @@
+  it "reports exceptions" do
+    Honeybadger.flush do
+      perform_enqueued_jobs do
+        post "/enqueue_error_job"
```
Is there a reason to have this run through a controller? Could it just enqueue the job directly?

```suggestion
        ErrorJob.perform_later({some: "data"})
```

halfbyte replied:
@joshuap Honestly, probably a combination of initial insecurity and maybe looking for examples in the wrong place. Looking at the code now, I'm not really sure, why I did that. I'll overhaul the test.

joshuap reviewed (commented):
This seems reasonable to me, assuming it works in a live Rails app. 😁

I left a question about the test harness.

joshuap commented on lib/honeybadger/plugins/active_job.rb:
```diff
@@ -0,0 +1,29 @@
+          ::ActiveJob::Base.class_eval do |base| 
+            base.set_callback :perform, :around do |param, block|
+              begin
+                Honeybadger.flush {
```
Using flush here means that any exceptions reported via `Honeybadger.notify` from inside the block call will be reported before continuing. The exception reported in the outer `rescue` statement on line 19 below will also be reported immediately because of the `sync: true` option.

subzero10 replied:
Is this the intended behavior though (to report immediately)?

halfbyte replied:
Good question. Not sure **I** can answer that ;). I think from what I've read most of the time the intention is to make testing easier but I guess wrapping the test code in `.flush {}` should be sufficient.

joshuap replied:
In general, `Honeybadger.flush` is necessary in cases where the process or thread could be killed before the agent finishes reporting error in the worker queue.

halfbyte commented:
@joshuap I've removed the :sync and the flush block, as I think both of them are not strictly necessary after reading your explanations. I should've looked at these options a bit more when copying the code over :)

I've also simplified the test and tested it in a rails app from the console.

joshuap reviewed (commented):
This is looking much better! Just one last question.

halfbyte commented on lib/honeybadger/plugins/active_job.rb:
```diff
@@ -0,0 +1,27 @@
+        execution {
+          ::ActiveJob::Base.class_eval do |base| 
+            base.set_callback :perform, :around do |param, block|
+              begin
```
@joshuap Thanks again for the explanations, I think this makes sense. The async adapater runs a `Concurrent::Threadpool`, so I assume the threads are persistent. I've added the call.

joshuap reviewed (approved):
:shipit:

halfbyte commented:
@subzero10 Can I merge this?

joshuap commented:
I'm good with merging this 👍

subzero10 commented:
Sure go ahead!
